In [ ]:
from scipy.io import arff
from sklearn.model_selection import train_test_split
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score
from keras import Sequential
from keras.layers import Dense
from keras.optimizers import Adam
import pandas as pd
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:80% !important; }</style>"))

In [1]:
data = arff.loadarff('rethinpathy.arff')
df = pd.DataFrame(data[0])

df.head()

0    1     2     3     4     5     6     7          8          9  \
0  1.0  1.0  22.0  22.0  22.0  19.0  18.0  14.0  49.895756  17.775994   
1  1.0  1.0  24.0  24.0  22.0  18.0  16.0  13.0  57.709936  23.799994   
2  1.0  1.0  62.0  60.0  59.0  54.0  47.0  33.0  55.831441  27.993933   
3  1.0  1.0  55.0  53.0  53.0  50.0  43.0  31.0  40.467228  18.445954   
4  1.0  1.0  44.0  44.0  44.0  41.0  39.0  27.0  18.026254   8.570709   

          10        11        12        13        14        15        16  \
0   5.270920  0.771761  0.018632  0.006864  0.003923  0.003923  0.486903   
1   3.325423  0.234185  0.003903  0.003903  0.003903  0.003903  0.520908   
2  12.687485  4.852282  1.393889  0.373252  0.041817  0.007744  0.530904   
3   9.118901  3.079428  0.840261  0.272434  0.007653  0.001531  0.483284   
4   0.410381  0.000000  0.000000  0.000000  0.000000  0.000000  0.475935   

         17   18 Class  
0  0.100025  1.0  b'0'  
1  0.144414  0.0  b'0'  
2  0.128548  0.0  b'1'  
3  0.114790  0.0  b'0'  
4  0.123572  0.0  b'1'

## Oznaczenie kolumn
0    - Oznaczenie jakości. 0 = zła jakość, 1 = dobra jakość<br>
1    - Wynik poprzedniego badania tj. czy było robione coś wczesniej, z czym pacjent przychodzi do okulisty. 1 oznacza nieprawidłowość siatkówki<br>
2-7  - Wynik wykrywania mikrotętniaków, wyrażone przez ilość takich poszerzeń naczynek<br>
8-15 - Znormalizowany wynik badania ilości wysięków. Normalizacja następuje przez iloczyn średnicy obszaru przez obraz który się bada<br>
16   - Znormalizowana odległość od środka plamki żółtej do środka tarczy nerwu wzrokowego <br>
17   - Średnica tarczy nerwu wzrokowego<br>
18   - Wyniki klasyfikacji patologicznych obrazów siatkówki za pomocą metody wykorzystującej modulacje amplitudy i częstotliwości<br>
19   - 1 = występują oznaki retinopatii, 0 = nie występują

In [2]:
df[df.columns[0:8]] = df[df.columns[0:8]].astype(int)
df[df.columns[18:20]] = df[df.columns[18:20]].astype(int)
col_names = ['Quality', 'Pre-screening',
             'MA result1', 'MA result2', 'MA result3', 'MA result4', 'MA result5', 'MA result6',
             'Exudates result1', 'Exudates result2', 'Exudates result3', 'Exudates result4', 'Exudates result5', 'Exudates result6', 'Exudates result7', 'Exudates result',
             'Macula - Optic disc dist', 'Optic disc diameter', 'AM/FM classification', 'Class']
df.columns = col_names
df.head()

Quality  Pre-screening  MA result1  MA result2  MA result3  MA result4  \
0        1              1          22          22          22          19   
1        1              1          24          24          22          18   
2        1              1          62          60          59          54   
3        1              1          55          53          53          50   
4        1              1          44          44          44          41   

   MA result5  MA result6  Exudates result1  Exudates result2  \
0          18          14         49.895756         17.775994   
1          16          13         57.709936         23.799994   
2          47          33         55.831441         27.993933   
3          43          31         40.467228         18.445954   
4          39          27         18.026254          8.570709   

   Exudates result3  Exudates result4  Exudates result5  Exudates result6  \
0          5.270920          0.771761          0.018632          0.006864   
1          3.325423          0.234185          0.003903          0.003903   
2         12.687485          4.852282          1.393889          0.373252   
3          9.118901          3.079428          0.840261          0.272434   
4          0.410381          0.000000          0.000000          0.000000   

   Exudates result7  Exudates result  Macula - Optic disc dist  \
0          0.003923         0.003923                  0.486903   
1          0.003903         0.003903                  0.520908   
2          0.041817         0.007744                  0.530904   
3          0.007653         0.001531                  0.483284   
4          0.000000         0.000000                  0.475935   

   Optic disc diameter  AM/FM classification  Class  
0             0.100025                     1      0  
1             0.144414                     0      0  
2             0.128548                     0      1  
3             0.114790                     0      0  
4             0.123572                     0      1

In [3]:
data = df[df.columns[0:19]]
target = df['Class']

col_names = ['Quality', 'Pre-screening', 'MA result1', 'MA result2', 'MA result3', 'MA result4', 'MA result5', 'MA result6', 'Exudates result1', 'Exudates result2', 'Exudates result3', 'Exudates result4', 'Exudates result5', 'Exudates result6', 'Exudates result7',
            'Exudates result', 'Macula - Optic disc dist', 'Optic disc diameter', 'AM/FM classification']
sc = StandardScaler()
data = sc.fit_transform(data)
data = pd.DataFrame(data, columns=col_names)

data_train, data_test, target_train, target_test = train_test_split(data,target, test_size = 0.20, random_state = 10)

In [4]:
#% chorych
target[target == 1].count() / target.count() * 100

53.08427454387489

In [5]:


svc_model = LinearSVC(random_state=0, max_iter = 10000000)

pred = svc_model.fit(data_train, target_train).predict(data_test)

print("LinearSVC acc: ", accuracy_score(target_test, pred, normalize = True))

LinearSVC acc:  0.7445887445887446


In [6]:
# neural network
adam = Adam(lr=0.01)

classifier = Sequential()
classifier.add(Dense(5, activation='sigmoid',
                     kernel_initializer='random_normal', input_dim=19))
classifier.add(Dense(5, activation='sigmoid',
                     kernel_initializer='random_normal'))
classifier.add(Dense(5, activation='sigmoid',
                     kernel_initializer='random_normal'))
classifier.add(Dense(1, activation='sigmoid',
                     kernel_initializer='random_normal'))
classifier.compile(
    optimizer=adam, loss='binary_crossentropy', metrics=['accuracy'])
classifier.fit(data_train, target_train, epochs=1000)

prednn = classifier.predict(data_test)

Using TensorFlow backend.


Epoch 1/1000
920/920 [==============================] - 0s 171us/step - loss: 0.6933 - accuracy: 0.4804
Epoch 2/1000
920/920 [==============================] - 0s 22us/step - loss: 0.6927 - accuracy: 0.5217
Epoch 3/1000
920/920 [==============================] - 0s 22us/step - loss: 0.6930 - accuracy: 0.5217
Epoch 4/1000
920/920 [==============================] - 0s 22us/step - loss: 0.6931 - accuracy: 0.5217
Epoch 5/1000
920/920 [==============================] - 0s 20us/step - loss: 0.6930 - accuracy: 0.5217
Epoch 6/1000
920/920 [==============================] - 0s 20us/step - loss: 0.6927 - accuracy: 0.5217
Epoch 7/1000
920/920 [==============================] - 0s 23us/step - loss: 0.6926 - accuracy: 0.5217
Epoch 8/1000
920/920 [==============================] - 0s 22us/step - loss: 0.6931 - accuracy: 0.5217
Epoch 9/1000
920/920 [==============================] - 0s 20us/step - loss: 0.6927 - accuracy: 0.5217
Epoch 10/1000
920/920 [==============================] - 0s 22us/step - 

920/920 [==============================] - 0s 18us/step - loss: 0.6312 - accuracy: 0.5957
Epoch 80/1000
920/920 [==============================] - 0s 18us/step - loss: 0.6021 - accuracy: 0.6457
Epoch 81/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5929 - accuracy: 0.6533
Epoch 82/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5888 - accuracy: 0.6739
Epoch 83/1000
920/920 [==============================] - 0s 20us/step - loss: 0.6053 - accuracy: 0.6663
Epoch 84/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5926 - accuracy: 0.6783
Epoch 85/1000
920/920 [==============================] - 0s 22us/step - loss: 0.5865 - accuracy: 0.6783
Epoch 86/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5909 - accuracy: 0.6772
Epoch 87/1000
920/920 [==============================] - 0s 22us/step - loss: 0.5794 - accuracy: 0.6902
Epoch 88/1000
920/920 [==============================] - 0s 21us/step - loss: 

920/920 [==============================] - 0s 20us/step - loss: 0.5643 - accuracy: 0.7043
Epoch 158/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5615 - accuracy: 0.6957
Epoch 159/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5738 - accuracy: 0.6837
Epoch 160/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5663 - accuracy: 0.6891
Epoch 161/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5387 - accuracy: 0.7152
Epoch 162/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5739 - accuracy: 0.7109
Epoch 163/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5659 - accuracy: 0.6902
Epoch 164/1000
920/920 [==============================] - 0s 21us/step - loss: 0.5507 - accuracy: 0.7076
Epoch 165/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5571 - accuracy: 0.7033
Epoch 166/1000
920/920 [==============================] - 0s 20us/step

920/920 [==============================] - 0s 18us/step - loss: 0.5371 - accuracy: 0.7250
Epoch 236/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5325 - accuracy: 0.7141
Epoch 237/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5301 - accuracy: 0.7250
Epoch 238/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5282 - accuracy: 0.7326
Epoch 239/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5180 - accuracy: 0.7359
Epoch 240/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5251 - accuracy: 0.7185
Epoch 241/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5221 - accuracy: 0.7239
Epoch 242/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5268 - accuracy: 0.7087
Epoch 243/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5229 - accuracy: 0.7272
Epoch 244/1000
920/920 [==============================] - 0s 19us/step

920/920 [==============================] - 0s 19us/step - loss: 0.5234 - accuracy: 0.7413
Epoch 314/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5135 - accuracy: 0.7337
Epoch 315/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5036 - accuracy: 0.7435
Epoch 316/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5190 - accuracy: 0.7272
Epoch 317/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5137 - accuracy: 0.7261
Epoch 318/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5137 - accuracy: 0.7185
Epoch 319/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5055 - accuracy: 0.7380
Epoch 320/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5305 - accuracy: 0.7283
Epoch 321/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5154 - accuracy: 0.7337
Epoch 322/1000
920/920 [==============================] - 0s 18us/step

920/920 [==============================] - 0s 17us/step - loss: 0.5102 - accuracy: 0.7478
Epoch 392/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5127 - accuracy: 0.7337
Epoch 393/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5140 - accuracy: 0.7207
Epoch 394/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5022 - accuracy: 0.7435
Epoch 395/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5265 - accuracy: 0.7163
Epoch 396/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5320 - accuracy: 0.7185
Epoch 397/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5065 - accuracy: 0.7424
Epoch 398/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5130 - accuracy: 0.7315
Epoch 399/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5023 - accuracy: 0.7359
Epoch 400/1000
920/920 [==============================] - 0s 17us/step

920/920 [==============================] - 0s 18us/step - loss: 0.5111 - accuracy: 0.7239
Epoch 469/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5067 - accuracy: 0.7489
Epoch 470/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5103 - accuracy: 0.7272
Epoch 471/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5045 - accuracy: 0.7359
Epoch 472/1000
920/920 [==============================] - 0s 19us/step - loss: 0.4965 - accuracy: 0.7630
Epoch 473/1000
920/920 [==============================] - 0s 19us/step - loss: 0.4980 - accuracy: 0.7391
Epoch 474/1000
920/920 [==============================] - 0s 20us/step - loss: 0.4978 - accuracy: 0.7424
Epoch 475/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5238 - accuracy: 0.7326
Epoch 476/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5185 - accuracy: 0.7250
Epoch 477/1000
920/920 [==============================] - 0s 18us/step

920/920 [==============================] - 0s 18us/step - loss: 0.5195 - accuracy: 0.7315
Epoch 547/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5070 - accuracy: 0.7446
Epoch 548/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5070 - accuracy: 0.7315
Epoch 549/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5070 - accuracy: 0.7174
Epoch 550/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5020 - accuracy: 0.7435
Epoch 551/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5447 - accuracy: 0.7011
Epoch 552/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5006 - accuracy: 0.7402
Epoch 553/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5057 - accuracy: 0.7402
Epoch 554/1000
920/920 [==============================] - 0s 17us/step - loss: 0.4948 - accuracy: 0.7435
Epoch 555/1000
920/920 [==============================] - 0s 17us/step

920/920 [==============================] - 0s 22us/step - loss: 0.4913 - accuracy: 0.7576
Epoch 625/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5014 - accuracy: 0.7511
Epoch 626/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5071 - accuracy: 0.7283
Epoch 627/1000
920/920 [==============================] - 0s 17us/step - loss: 0.4955 - accuracy: 0.7543
Epoch 628/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4949 - accuracy: 0.7478
Epoch 629/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5049 - accuracy: 0.7413
Epoch 630/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4879 - accuracy: 0.7565
Epoch 631/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5367 - accuracy: 0.7250
Epoch 632/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5146 - accuracy: 0.7489
Epoch 633/1000
920/920 [==============================] - 0s 17us/step

920/920 [==============================] - 0s 17us/step - loss: 0.5111 - accuracy: 0.7402
Epoch 703/1000
920/920 [==============================] - 0s 17us/step - loss: 0.4997 - accuracy: 0.7522
Epoch 704/1000
920/920 [==============================] - 0s 17us/step - loss: 0.5282 - accuracy: 0.7228
Epoch 705/1000
920/920 [==============================] - 0s 19us/step - loss: 0.4951 - accuracy: 0.7457
Epoch 706/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4973 - accuracy: 0.7565
Epoch 707/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5026 - accuracy: 0.7467
Epoch 708/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5058 - accuracy: 0.7489
Epoch 709/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5050 - accuracy: 0.7326
Epoch 710/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5126 - accuracy: 0.7337
Epoch 711/1000
920/920 [==============================] - 0s 17us/step

920/920 [==============================] - 0s 20us/step - loss: 0.4910 - accuracy: 0.7576
Epoch 781/1000
920/920 [==============================] - 0s 19us/step - loss: 0.4903 - accuracy: 0.7739
Epoch 782/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5023 - accuracy: 0.7489
Epoch 783/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5109 - accuracy: 0.7402
Epoch 784/1000
920/920 [==============================] - 0s 20us/step - loss: 0.4997 - accuracy: 0.7424
Epoch 785/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5039 - accuracy: 0.7380
Epoch 786/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5074 - accuracy: 0.7402
Epoch 787/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4942 - accuracy: 0.7565
Epoch 788/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4873 - accuracy: 0.7565
Epoch 789/1000
920/920 [==============================] - 0s 18us/step

920/920 [==============================] - 0s 20us/step - loss: 0.5143 - accuracy: 0.7239
Epoch 859/1000
920/920 [==============================] - 0s 19us/step - loss: 0.4916 - accuracy: 0.7576
Epoch 860/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4882 - accuracy: 0.7467
Epoch 861/1000
920/920 [==============================] - 0s 20us/step - loss: 0.4956 - accuracy: 0.7489
Epoch 862/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5064 - accuracy: 0.7348
Epoch 863/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4916 - accuracy: 0.7446
Epoch 864/1000
920/920 [==============================] - 0s 20us/step - loss: 0.4932 - accuracy: 0.7522
Epoch 865/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5071 - accuracy: 0.7326
Epoch 866/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5040 - accuracy: 0.7467
Epoch 867/1000
920/920 [==============================] - 0s 19us/step

920/920 [==============================] - 0s 17us/step - loss: 0.4917 - accuracy: 0.7587
Epoch 937/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4957 - accuracy: 0.7620
Epoch 938/1000
920/920 [==============================] - 0s 19us/step - loss: 0.4981 - accuracy: 0.7446
Epoch 939/1000
920/920 [==============================] - 0s 18us/step - loss: 0.5017 - accuracy: 0.7272
Epoch 940/1000
920/920 [==============================] - 0s 20us/step - loss: 0.5134 - accuracy: 0.7380
Epoch 941/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5148 - accuracy: 0.7391
Epoch 942/1000
920/920 [==============================] - 0s 17us/step - loss: 0.4930 - accuracy: 0.7620
Epoch 943/1000
920/920 [==============================] - 0s 18us/step - loss: 0.4895 - accuracy: 0.7543
Epoch 944/1000
920/920 [==============================] - 0s 19us/step - loss: 0.5023 - accuracy: 0.7424
Epoch 945/1000
920/920 [==============================] - 0s 17us/step

In [7]:
prednn = (prednn > 0.5)
print("NN accuracy: ", accuracy_score(target_test, prednn))

NN accuracy:  0.7489177489177489


## Podsumowanie
Prosta klasyfikacja została przeprowadzona na zbiorze rozmiarze około 1000 rekordów zawierających dane na temat badań pacjentów pod kątem retinopatii cukrzucowej. W zbiorze około 54% pacjentów było chorych. Załadowany dataframe zawiera 19 kolumn z wynikami badań oraz jedną klasyfikującą czy występuje stan patologiczny. Dane zostały podzielone w stosunku 80% dane trenujące i 20% dane testowe. Przeprowadzono testy za pomocą następujących algorytmów: Naive Bayes, LinearSVC, KNN, Decision Tree AdaBoost oraz z wykorzystaniem prostej sieci neuronowej. Najlepsze wyniki zostały osiągnięte dla LinearSVC oraz dla sieci neuronowych - około 74% accuracy. W tym pliku zostały pominiete klasyfikatory o słabszych wynikach. 